## Assignment 2 - Part 2
* Topics: pandas and numpy

In [18]:
from sklearn.datasets import load_iris
import pandas as pd
import numpy as np

iris = load_iris()

In [74]:
feature_names = iris.feature_names
iris_dataframe = pd.DataFrame(iris.data, columns = feature_names)
#target = pd.DataFrame(iris.target, columns = ['target'])
target = pd.Series(iris.target, name='flower')
target_names = iris.target_names


In [76]:
iris_dataframe['target'] = np.where(target == 0, 'setosa', (np.where(target == 1, 'versicolor', 'virginica')))


In [77]:
iris_dataframe

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

        target  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 5 columns]

## Question 6
* sort the feature values for each row and replace the indices with the feature names
    * so each row will have 4 columns, the first column being the feature name that is the highest value, the second column being the feature name that is the second highest value, etc.
* note, watch out how argsort in numpy works.  you will need to reverse the order somehow.  the sorted(reverse = True) functionality might be of help.
* make sure to replace the index values with the feature name, as we did above
* put the resulting 2-d array into a pandas dataframe and print the first 5 rows.
* hint, look at the apply_along_axis() method for numpy arrays

In [26]:
feature_names = iris.feature_names

np.sort(iris.data).sort()

sorted_iris = iris.data.argsort()

In [52]:
sorted_array = np.where(sorted_iris == 0, feature_names[0], np.where(sorted_iris == 1, feature_names[1], np.where(sorted_iris == 2, feature_names[2], feature_names[3])))

In [57]:
my_answer = pd.DataFrame(sorted_array,columns=['Highest','High', 'Low', 'Lowest'])

In [58]:
my_answer.head()

Highest               High               Low             Lowest
0  petal width (cm)  petal length (cm)  sepal width (cm)  sepal length (cm)
1  petal width (cm)  petal length (cm)  sepal width (cm)  sepal length (cm)
2  petal width (cm)  petal length (cm)  sepal width (cm)  sepal length (cm)
3  petal width (cm)  petal length (cm)  sepal width (cm)  sepal length (cm)
4  petal width (cm)  petal length (cm)  sepal width (cm)  sepal length (cm)

## Question 7
* apply z-score normalization to each column in the iris data (note you do not need the target/label column)
* column wise meaning, treat each column as an array, and find the standard deviations and means of that column
* note, you can use zscore from scipy.stats
* results should be a pandas dataframe, printing out 5 rows and running the describe() method on the dataframe

In [7]:
mu = iris_dataframe.iloc[:,0:4].mean()
sd = iris_dataframe.iloc[:,0:4].var()**.5

In [12]:
standerdize_iris = (iris_dataframe.iloc[:,0:4]-mu)/sd

In [15]:
standerdize_iris.iloc[0:5,]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0          -0.897674          1.015602          -1.335752         -1.311052
1          -1.139200         -0.131539          -1.335752         -1.311052
2          -1.380727          0.327318          -1.392399         -1.311052
3          -1.501490          0.097889          -1.279104         -1.311052
4          -1.018437          1.245030          -1.335752         -1.311052

In [14]:
standerdize_iris.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count       1.500000e+02      1.500000e+02       1.500000e+02   
mean       -1.457168e-15     -1.638319e-15      -1.292300e-15   
std         1.000000e+00      1.000000e+00       1.000000e+00   
min        -1.863780e+00     -2.425820e+00      -1.562342e+00   
25%        -8.976739e-01     -5.903951e-01      -1.222456e+00   
50%        -5.233076e-02     -1.315388e-01       3.353541e-01   
75%         6.722490e-01      5.567457e-01       7.602115e-01   
max         2.483699e+00      3.080455e+00       1.779869e+00   

       petal width (cm)  
count      1.500000e+02  
mean      -5.543714e-16  
std        1.000000e+00  
min       -1.442245e+00  
25%       -1.179859e+00  
50%        1.320673e-01  
75%        7.880307e-01  
max        1.706379e+00

## Question 8
* make a function that takes in a 2d numpy array X, a 1d numpy array Y and the size the training data, test data and valiation datasets.  
* return 6 items
    * train_x, train_y, test_x, test_y, val_x, val_y
* do not use any modules.  this can be solved using bracket notation to subset.  Note // will take care of decimals in doing division. you could also use int() to convert the float to a whole number
* make the params for the training, test and val size be decimals, that repsent percentages of the data.  For instance .8, .1, .1 means an 80% training, 10% test and 10% validation split
* use an assert to make sure the numbers add to 1
* print out the dimensions of all 6 elements, using iris as a test case with an 80-10-10 split

In [109]:
def train_test_validate(x, y, train_size = .8, test_size = .1, validate_size = .1):
    assert (train_size+test_size+validate_size == 1), 'data not split correctly, must add to 1. Use decimals.'
    
    index = set(range(0,len(x)))
    data_len = len(x)
    
    import random
    train_samples = random.sample(index,int(train_size*data_len))
    [index.remove(i) for i in train_samples]
    
    train_x = x[train_samples]
    train_y = y[train_samples]
    
    test_samples = random.sample(index,int(test_size*data_len))
    [index.remove(i) for i in test_samples]
    
    test_x = x[test_samples]
    test_y = y[test_samples]
    
    validate_samples = list(index)
    
    val_x = x[validate_samples]
    val_y = y[validate_samples]
    
    return train_x, train_y, test_x, test_y, val_x, val_y

In [110]:
x = iris_dataframe.iloc[:, 0:4].values
y = iris_dataframe.target.values

In [111]:
train_x, train_y, test_x, test_y, val_x, val_y = train_test_validate(x,y,train_size = .8, test_size = .1, validate_size = .1)

In [112]:
print(train_x, train_y, test_x, test_y, val_x, val_y)

[[6.5 3.2 5.1 2. ]
 [5.3 3.7 1.5 0.2]
 [5.6 2.5 3.9 1.1]
 [5.1 2.5 3.  1.1]
 [6.  2.7 5.1 1.6]
 [6.2 2.8 4.8 1.8]
 [5.8 2.7 5.1 1.9]
 [6.5 2.8 4.6 1.5]
 [5.4 3.7 1.5 0.2]
 [7.  3.2 4.7 1.4]
 [4.9 2.4 3.3 1. ]
 [5.4 3.  4.5 1.5]
 [6.1 2.6 5.6 1.4]
 [5.7 2.9 4.2 1.3]
 [5.5 2.4 3.7 1. ]
 [6.4 2.9 4.3 1.3]
 [5.  3.5 1.6 0.6]
 [5.6 3.  4.5 1.5]
 [6.  3.4 4.5 1.6]
 [5.7 3.8 1.7 0.3]
 [4.7 3.2 1.6 0.2]
 [5.5 4.2 1.4 0.2]
 [6.3 2.5 5.  1.9]
 [4.5 2.3 1.3 0.3]
 [6.1 2.8 4.7 1.2]
 [5.  3.4 1.6 0.4]
 [4.9 3.1 1.5 0.2]
 [6.7 3.3 5.7 2.1]
 [5.6 2.8 4.9 2. ]
 [4.6 3.6 1.  0.2]
 [6.9 3.1 4.9 1.5]
 [5.9 3.  5.1 1.8]
 [7.2 3.2 6.  1.8]
 [7.3 2.9 6.3 1.8]
 [5.8 2.7 4.1 1. ]
 [4.8 3.4 1.6 0.2]
 [5.  2.3 3.3 1. ]
 [5.4 3.9 1.7 0.4]
 [6.1 3.  4.9 1.8]
 [5.5 2.4 3.8 1.1]
 [6.7 3.3 5.7 2.5]
 [6.  3.  4.8 1.8]
 [4.6 3.2 1.4 0.2]
 [5.8 2.6 4.  1.2]
 [5.7 2.6 3.5 1. ]
 [6.3 2.7 4.9 1.8]
 [6.5 3.  5.8 2.2]
 [4.7 3.2 1.3 0.2]
 [6.7 3.1 4.7 1.5]
 [5.6 2.7 4.2 1.3]
 [4.9 3.6 1.4 0.1]
 [6.2 2.9 4.3 1.3]
 [6.3 2.3 4.

## Question 9
* using pandas, find the sum of each feature by species type (label)
* do the same, but find the min, max and median as well

In [78]:
iris_dataframe.groupby('target').sum()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
target                                                               
setosa                  250.3             171.4               73.1   
versicolor              296.8             138.5              213.0   
virginica               329.4             148.7              277.6   

            petal width (cm)  
target                        
setosa                  12.3  
versicolor              66.3  
virginica              101.3

In [7]:
iris_dataframe.groupby('target').max()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
target                                                               
setosa                    5.8               4.4                1.9   
versicolor                7.0               3.4                5.1   
virginica                 7.9               3.8                6.9   

            petal width (cm)  
target                        
setosa                   0.6  
versicolor               1.8  
virginica                2.5

In [8]:
iris_dataframe.groupby('target').min()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
target                                                               
setosa                    4.3               2.3                1.0   
versicolor                4.9               2.0                3.0   
virginica                 4.9               2.2                4.5   

            petal width (cm)  
target                        
setosa                   0.1  
versicolor               1.0  
virginica                1.4

In [95]:
iris_dataframe.groupby('target').median()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
target                                                               
setosa                    5.0               3.4               1.50   
versicolor                5.9               2.8               4.35   
virginica                 6.5               3.0               5.55   

            petal width (cm)  
target                        
setosa                   0.2  
versicolor               1.3  
virginica                2.0

## Question 10
* mean center each column of the iris dataframe (excluding the label column of course)
* this means, for each column, the mean should be zero.  to accomplish this we can subtract the column mean from each element of the column
* note, broadcasting, which if we have say a 150 row and 4 column dataframe, can  take a row vector of size 4 and apply it to each element
*  thus, we can answer this questions doing something like df - df_col_means
* run the describe() method at the end to show the data has been mean center

In [75]:
#get mean of each column
names_and_means = iris_dataframe.mean()

In [55]:
#store name of each column in generator
col_names = (k for k in names_and_means.index)

#store column mean values in generator
col_mean = (k for k in names_and_means)

In [56]:
for col_number in range(0,4):
    x = next(col_names)
    iris_dataframe[x] = iris_dataframe[x] - next(col_mean)

In [57]:
iris_dataframe.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count       1.500000e+02      1.500000e+02       1.500000e+02   
mean       -1.109298e-16     -1.603347e-16      -4.292862e-16   
std         8.280661e-01      4.358663e-01       1.765298e+00   
min        -1.543333e+00     -1.057333e+00      -2.758000e+00   
25%        -7.433333e-01     -2.573333e-01      -2.158000e+00   
50%        -4.333333e-02     -5.733333e-02       5.920000e-01   
75%         5.566667e-01      2.426667e-01       1.342000e+00   
max         2.056667e+00      1.342667e+00       3.142000e+00   

       petal width (cm)  
count      1.500000e+02  
mean      -5.625130e-17  
std        7.622377e-01  
min       -1.099333e+00  
25%       -8.993333e-01  
50%        1.006667e-01  
75%        6.006667e-01  
max        1.300667e+00

In [87]:
#another solution. Be sure to re-set Iris dataframe
a,b,c,d = iris_dataframe.mean()

In [88]:
newDF = iris_dataframe.iloc[:,0:4]-[a,b,c,d]

In [91]:
pd.DataFrame(newDF).describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count       1.500000e+02      1.500000e+02       1.500000e+02   
mean       -1.125026e-15     -7.608728e-16      -2.552033e-15   
std         8.280661e-01      4.358663e-01       1.765298e+00   
min        -1.543333e+00     -1.057333e+00      -2.758000e+00   
25%        -7.433333e-01     -2.573333e-01      -2.158000e+00   
50%        -4.333333e-02     -5.733333e-02       5.920000e-01   
75%         5.566667e-01      2.426667e-01       1.342000e+00   
max         2.056667e+00      1.342667e+00       3.142000e+00   

       petal width (cm)  
count      1.500000e+02  
mean      -4.485301e-16  
std        7.622377e-01  
min       -1.099333e+00  
25%       -8.993333e-01  
50%        1.006667e-01  
75%        6.006667e-01  
max        1.300667e+00

## Extra questions.. to think and answer..
* Explain what the axis mean in regards to a pandas dataframe and numpy array
* How would you groupby two columns in pandas?
* What functions are used to read in csvs and excel files in pandas

Since pandas is built on top of numpy, the axis of both refers to which dimmenson of the data set we are looking at. For example, axis 0 goes top down. If we were to sum on axis 0, the data collapses the rows into one row and sums the columns. Axis 1 will collapse the columns into a singe column and sum the rows. 

In [84]:
np.sum(iris.data, axis = 0)

array([876.5, 458.6, 563.7, 179.9])

In [85]:
np.sum(iris.data, axis = 1)

array([10.2,  9.5,  9.4,  9.4, 10.2, 11.4,  9.7, 10.1,  8.9,  9.6, 10.8,
       10. ,  9.3,  8.5, 11.2, 12. , 11. , 10.3, 11.5, 10.7, 10.7, 10.7,
        9.4, 10.6, 10.3,  9.8, 10.4, 10.4, 10.2,  9.7,  9.7, 10.7, 10.9,
       11.3,  9.7,  9.6, 10.5, 10. ,  8.9, 10.2, 10.1,  8.4,  9.1, 10.7,
       11.2,  9.5, 10.7,  9.4, 10.7,  9.9, 16.3, 15.6, 16.4, 13.1, 15.4,
       14.3, 15.9, 11.6, 15.4, 13.2, 11.5, 14.6, 13.2, 15.1, 13.4, 15.6,
       14.6, 13.6, 14.4, 13.1, 15.7, 14.2, 15.2, 14.8, 14.9, 15.4, 15.8,
       16.4, 14.9, 12.8, 12.8, 12.6, 13.6, 15.4, 14.4, 15.5, 16. , 14.3,
       14. , 13.3, 13.7, 15.1, 13.6, 11.6, 13.8, 14.1, 14.1, 14.7, 11.7,
       13.9, 18.1, 15.5, 18.1, 16.6, 17.5, 19.3, 13.6, 18.3, 16.8, 19.4,
       16.8, 16.3, 17.4, 15.2, 16.1, 17.2, 16.8, 20.4, 19.5, 14.7, 18.1,
       15.3, 19.2, 15.7, 17.8, 18.2, 15.6, 15.8, 16.9, 17.6, 18.2, 20.1,
       17. , 15.7, 15.7, 19.1, 17.7, 16.8, 15.6, 17.5, 17.8, 17.4, 15.5,
       18.2, 18.2, 17.2, 15.7, 16.7, 17.3, 15.8])

In [91]:
iris_dataframe.iloc[:,0:4].sum(axis = 0)

sepal length (cm)    876.5
sepal width (cm)     458.6
petal length (cm)    563.7
petal width (cm)     179.9
dtype: float64

In [88]:
iris_dataframe.sum(axis = 1)

0      10.2
1       9.5
2       9.4
3       9.4
4      10.2
       ... 
145    17.2
146    15.7
147    16.7
148    17.3
149    15.8
Length: 150, dtype: float64

To group by two columns, add a list of columns you want to group by to function. Iris is not best example, but here you go:

In [97]:
iris_dataframe.groupby(['target', 'sepal width (cm)']).median()

sepal length (cm)  petal length (cm)  \
target     sepal width (cm)                                         
setosa     2.3                            4.50               1.30   
           2.9                            4.40               1.40   
           3.0                            4.80               1.40   
           3.1                            4.85               1.50   
           3.2                            4.70               1.30   
           3.3                            5.05               1.55   
           3.4                            5.00               1.50   
           3.5                            5.10               1.40   
           3.6                            4.90               1.40   
           3.7                            5.30               1.50   
           3.8                            5.10               1.65   
           3.9                            5.40               1.50   
           4.0                            5.80               1.20   
           4.1                            5.20               1.50   
           4.2                            5.50               1.40   
           4.4                            5.70               1.50   
versicolor 2.0                            5.00               3.50   
           2.2                            6.10               4.25   
           2.3                            5.50               4.00   
           2.4                            5.50               3.70   
           2.5                            5.55               3.95   
           2.6                            5.70               4.00   
           2.7                            5.80               4.10   
           2.8                            6.10               4.55   
           2.9                            6.10               4.30   
           3.0                            5.80               4.45   
           3.1                            6.70               4.70   
           3.2                            6.40               4.70   
           3.3                            6.30               4.70   
           3.4                            6.00               4.50   
virginica  2.2                            6.00               5.00   
           2.5                            6.00               5.00   
           2.6                            6.90               6.25   
           2.7                            6.05               5.10   
           2.8                            6.35               5.35   
           2.9                            6.80               5.95   
           3.0                            6.60               5.50   
           3.1                            6.80               5.45   
           3.2                            6.80               5.70   
           3.3                            6.70               5.70   
           3.4                            6.25               5.50   
           3.6                            7.20               6.10   
           3.8                            7.80               6.55   

                             petal width (cm)  
target     sepal width (cm)                    
setosa     2.3                           0.30  
           2.9                           0.20  
           3.0                           0.20  
           3.1                           0.20  
           3.2                           0.20  
           3.3                           0.35  
           3.4                           0.20  
           3.5                           0.25  
           3.6                           0.20  
           3.7                           0.20  
           3.8                           0.30  
           3.9                           0.40  
           4.0                           0.20  
           4.1                           0.10  
           4.2                           0.20  
           4.4                           0.40  
versicolor 2.0                           1.00  
           

pd.read_csv is the function to read .csv in pandas.
pd.read_exce is the function to read .xlsx in pandas.